<a href="https://colab.research.google.com/github/sk1252k5/deeplearning_TF_Keras/blob/main/Droupout_Layer/Dropout_layer.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

In [1]:
# importing the dataset
import tensorflow as tf
from tensorflow import keras
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt

In [2]:
# loading the dataset
df = pd.read_csv('/content/sonar_dataset.csv',header = None)
df.head(5)

0       1       2       3       4       5       6       7       8   \
0  0.0200  0.0371  0.0428  0.0207  0.0954  0.0986  0.1539  0.1601  0.3109   
1  0.0453  0.0523  0.0843  0.0689  0.1183  0.2583  0.2156  0.3481  0.3337   
2  0.0262  0.0582  0.1099  0.1083  0.0974  0.2280  0.2431  0.3771  0.5598   
3  0.0100  0.0171  0.0623  0.0205  0.0205  0.0368  0.1098  0.1276  0.0598   
4  0.0762  0.0666  0.0481  0.0394  0.0590  0.0649  0.1209  0.2467  0.3564   

       9   ...      51      52      53      54      55      56      57  \
0  0.2111  ...  0.0027  0.0065  0.0159  0.0072  0.0167  0.0180  0.0084   
1  0.2872  ...  0.0084  0.0089  0.0048  0.0094  0.0191  0.0140  0.0049   
2  0.6194  ...  0.0232  0.0166  0.0095  0.0180  0.0244  0.0316  0.0164   
3  0.1264  ...  0.0121  0.0036  0.0150  0.0085  0.0073  0.0050  0.0044   
4  0.4459  ...  0.0031  0.0054  0.0105  0.0110  0.0015  0.0072  0.0048   

       58      59  60  
0  0.0090  0.0032   R  
1  0.0052  0.0044   R  
2  0.0095  0.0078   R  
3  0.0040  0.0117   R  
4  0.0107  0.0094   R  

[5 rows x 61 columns]

In [3]:
df.shape

(208, 61)

In [4]:
df[60].value_counts()

60
M    111
R     97
Name: count, dtype: int64

In [5]:
x = df.drop(60, axis=1)
y = df[60]
y.head()

0    R
1    R
2    R
3    R
4    R
Name: 60, dtype: object

In [6]:
y = pd.get_dummies(y,drop_first=True)
y = y.astype(int)
y.head(5)

R
0  1
1  1
2  1
3  1
4  1

In [7]:
y.dtypes

R    int64
dtype: object

In [8]:
x.shape

(208, 60)

In [10]:
from sklearn.model_selection import train_test_split
X_train, X_test, y_train, y_test = train_test_split(x, y, test_size=0.25, random_state=1)

In [11]:
# building the neural network
model = keras.Sequential([
    keras.layers.Dense(60,input_shape=(60,),activation='relu'),
    keras.layers.Dense(30,activation='relu'),
    keras.layers.Dense(15,activation='relu'),
    keras.layers.Dense(1,activation='sigmoid')
])
model.compile(optimizer='adam',loss='binary_crossentropy',metrics=['accuracy'])
model.fit(X_train,y_train,epochs=100,batch_size=8)

Epoch 1/100
20/20 [==============================] - 3s 8ms/step - loss: 0.6901 - accuracy: 0.5385
Epoch 2/100
20/20 [==============================] - 0s 5ms/step - loss: 0.6653 - accuracy: 0.6603
Epoch 3/100
20/20 [==============================] - 0s 6ms/step - loss: 0.6431 - accuracy: 0.6026
Epoch 4/100
20/20 [==============================] - 0s 8ms/step - loss: 0.6207 - accuracy: 0.7500
Epoch 5/100
20/20 [==============================] - 0s 10ms/step - loss: 0.5831 - accuracy: 0.7564
Epoch 6/100
20/20 [==============================] - 0s 5ms/step - loss: 0.5512 - accuracy: 0.8013
Epoch 7/100
20/20 [==============================] - 0s 7ms/step - loss: 0.5180 - accuracy: 0.7564
Epoch 8/100
20/20 [==============================] - 0s 7ms/step - loss: 0.4711 - accuracy: 0.8333
Epoch 9/100
20/20 [==============================] - 0s 5ms/step - loss: 0.4580 - accuracy: 0.7628
Epoch 10/100
20/20 [==============================] - 0s 7ms/step - loss: 0.4309 - accuracy: 0.7949
Epoch 11

In [12]:
model.evaluate(X_test,y_test)

2/2 [==============================] - 0s 9ms/step - loss: 1.2078 - accuracy: 0.7500


[1.2078430652618408, 0.75]

In [13]:
y_pred = model.predict(X_test).reshape(-1)
y_pred[:10]

2/2 [==============================] - 0s 6ms/step


array([1.12423173e-07, 9.98598874e-01, 9.93881047e-01, 1.08745835e-05,
       9.99999940e-01, 9.99931753e-01, 2.92626441e-01, 1.00000000e+00,
       3.41920622e-06, 1.00000000e+00], dtype=float32)

In [14]:
y_test[:10]

R
186  0
155  0
165  0
200  0
58   1
34   1
151  0
18   1
202  0
62   1

In [15]:
for i in range(len(y_pred)):
  if y_pred[i]>0.5:
    y_pred[i]=1
  else:
    y_pred[i]=0

In [16]:
y_pred[:10]

array([0., 1., 1., 0., 1., 1., 0., 1., 0., 1.], dtype=float32)

In [17]:
from sklearn.metrics import confusion_matrix , classification_report

print(classification_report(y_test, y_pred))

              precision    recall  f1-score   support

           0       0.72      0.85      0.78        27
           1       0.80      0.64      0.71        25

    accuracy                           0.75        52
   macro avg       0.76      0.75      0.75        52
weighted avg       0.76      0.75      0.75        52



**Model with Dropout Layer**

In [18]:
# building the neural network
model = keras.Sequential([
    keras.layers.Dense(60,input_shape=(60,),activation='relu'),
    keras.layers.Dropout(0.5),
    keras.layers.Dense(30,activation='relu'),
    keras.layers.Dropout(0.5),
    keras.layers.Dense(15,activation='relu'),
    keras.layers.Dropout(0.5),
    keras.layers.Dense(1,activation='sigmoid')
])
model.compile(optimizer='adam',loss='binary_crossentropy',metrics=['accuracy'])
model.fit(X_train,y_train,epochs=100,batch_size=8)

Epoch 1/100
20/20 [==============================] - 2s 5ms/step - loss: 0.7040 - accuracy: 0.5000
Epoch 2/100
20/20 [==============================] - 0s 4ms/step - loss: 0.7268 - accuracy: 0.4359
Epoch 3/100
20/20 [==============================] - 0s 3ms/step - loss: 0.7188 - accuracy: 0.5192
Epoch 4/100
20/20 [==============================] - 0s 3ms/step - loss: 0.6919 - accuracy: 0.4808
Epoch 5/100
20/20 [==============================] - 0s 3ms/step - loss: 0.6984 - accuracy: 0.4744
Epoch 6/100
20/20 [==============================] - 0s 2ms/step - loss: 0.6898 - accuracy: 0.5128
Epoch 7/100
20/20 [==============================] - 0s 3ms/step - loss: 0.6861 - accuracy: 0.5128
Epoch 8/100
20/20 [==============================] - 0s 2ms/step - loss: 0.7000 - accuracy: 0.5128
Epoch 9/100
20/20 [==============================] - 0s 3ms/step - loss: 0.6823 - accuracy: 0.5577
Epoch 10/100
20/20 [==============================] - 0s 2ms/step - loss: 0.6937 - accuracy: 0.5385
Epoch 11/

In [19]:
model.evaluate(X_test,y_test)

2/2 [==============================] - 0s 10ms/step - loss: 0.4764 - accuracy: 0.7500


[0.4764086604118347, 0.75]

In [20]:
y_pred = model.predict(X_test).reshape(-1)
for i in range(len(y_pred)):
  if y_pred[i]>0.5:
    y_pred[i]=1
  else:
    y_pred[i]=0
y_pred[:10]

2/2 [==============================] - 0s 6ms/step


array([0., 1., 1., 0., 1., 1., 0., 1., 0., 1.], dtype=float32)

In [21]:
y_test[:10]

R
186  0
155  0
165  0
200  0
58   1
34   1
151  0
18   1
202  0
62   1

In [22]:
from sklearn.metrics import confusion_matrix , classification_report

print(classification_report(y_test, y_pred))

              precision    recall  f1-score   support

           0       0.71      0.89      0.79        27
           1       0.83      0.60      0.70        25

    accuracy                           0.75        52
   macro avg       0.77      0.74      0.74        52
weighted avg       0.77      0.75      0.74        52



You can see that by using dropout layer test accuracy increased from 0.77 to 0.81